[JIT] implement tuple constants

eellison opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#31841 [JIT] implement tuple constants**
* #31840 [JIT] remove tuple logic in constant propagation
* #31839 [JIT] improve mayContainAlias
* #31791 Add str[] float[] constants resubmit

Add Tuple Constants to JIT. The constraint here is that all elements of a tuple must themself be insertable as a a constant. Previously tuples were special cased in constant propagation, but now that there are more passes that are inserted constants, such as freezing, we should just have tuples be representable as constants.

Differential Revision: [D19439514](https://our.internmc.facebook.com/intern/diff/D19439514)

Commit:
implement tuple constants

[ghstack-poisoned]

kostmo commented:
## :pill: CircleCI build failures summary and remediations
As of commit 7af9aa0a:


* **1/1** failures introduced in this PR


---

<details><summary>This comment was automatically generated by <a href="https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment">Dr. CI</a> (expand for details).</summary>Follow <a href="https://dr.pytorch.org/admin/comments-opt-out.html">this link to opt-out</a> of these comments for your Pull Requests.

Please report bugs/suggestions on the <a href="https://github.com/kostmo/circleci-failure-tracker/issues">GitHub issue tracker</a>.
</details>

zdevito commented on torch/csrc/jit/attributes.h:
```diff
@@ -19,10 +19,21 @@ using ::c10::Symbol;
 constexpr int max_tensor_display_size = 10;
 
-enum class AttributeKind { f, fs, i, is, s, ss, t, ts, g, gs, ty, tys };
+enum class AttributeKind { f, fs, i, is, s, ss, t, ts, g, gs, ty, tys, tup };
```
I have concerns about the gradient of this change. Every additional value that needs to be a constant will now need to be re-encoded as a attribute in the ir. As this expands, IR will reproduce the functionality of IValue. If we were starting from scratch, the _only_ attribute type a Node would have would be an IValue. We can't easily do that now, because too many downstream things, like onnx, use attributes. However, I think it is reasonable to add IValue as an attribute type and use it for tuple constants. The way I see this going is:

1. This Patch: adds IValue as an attribute type, along with support for using it to hold Tuple constant. Should be a small change to this patch because handling of other constants remain the same.
2. IValues gain a repr (i.e. how do I serialize this thing as python text, so that python_print can read it). Switch uses of quoted strings to use repr as well and clean up python_print.
3. Store  f, fs, i, is, s, ss as IValues as well. So that we do not change the externally visible API, the f, f_ functions should remain, but they should just do a type check that the underlying IValue is a float (or int, or string). Constants should now work by just directly storing the IValue.

eellison replied:
So, as far as this patch goes, what is the concrete difference ? instead of `TupleAttr` there is `IValueAttr`, and the `tup_` accessor converts the Tuple IValue into the general IValueAttr ?

As far as repr: that seems like a good change, containers such as tuple will still need to be special cased for python print because of the special `CONSTANTS.0` exporting of tuple constants

Commit:
Update on "[JIT] implement tuple constants"


Add Tuple Constants to JIT. The constraint here is that all elements of a tuple must themself be insertable as a a constant. Previously tuples were special cased in constant propagation, but now that there are more passes that are inserted constants, such as freezing, we should just have tuples be representable as constants. 



[ghstack-poisoned]

Commit:
Update on "[JIT] implement tuple constants"


Add Tuple Constants to JIT. The constraint here is that all elements of a tuple must themself be insertable as a a constant. Previously tuples were special cased in constant propagation, but now that there are more passes that are inserted constants, such as freezing, we should just have tuples be representable as constants. 



[ghstack-poisoned]

Commit:
Update on "[JIT] implement tuple constants"


Add Tuple Constants to JIT. The constraint here is that all elements of a tuple must themself be insertable as a a constant. Previously tuples were special cased in constant propagation, but now that there are more passes that are inserted constants, such as freezing, we should just have tuples be representable as constants. 



[ghstack-poisoned]

Commit:
Update on "[JIT] implement tuple constants"


Add Tuple Constants to JIT. The constraint here is that all elements of a tuple must themself be insertable as a a constant. Previously tuples were special cased in constant propagation, but now that there are more passes that are inserted constants, such as freezing, we should just have tuples be representable as constants. 



[ghstack-poisoned]

Commit:
Update on "[JIT] implement tuple constants"


Add Tuple Constants to JIT. The constraint here is that all elements of a tuple must themself be insertable as a a constant. Previously tuples were special cased in constant propagation, but now that there are more passes that are inserted constants, such as freezing, we should just have tuples be representable as constants. 



[ghstack-poisoned]

eellison commented:
@zdevito i rebased past the PR removing list specialization. Review plz ?

Commit:
Update on "[JIT] implement tuple constants"


Add Tuple Constants to JIT. The constraint here is that all elements of a tuple must themself be insertable as a a constant. Previously tuples were special cased in constant propagation, but now that there are more passes that are inserted constants, such as freezing, we should just have tuples be representable as constants. 



[ghstack-poisoned]

zdevito commented on torch/csrc/jit/passes/python_print.cpp:
```diff
@@ -834,6 +835,19 @@ struct PythonPrintImpl {
       }
+    } else if (v.isTuple()) {
+      const auto& elems = v.toTuple()->elements();
+      ss << "(";
```
Can we deduplicate the logic for printing a tuple with the TupleConstruct printing?

zdevito commented on torch/csrc/jit/passes/onnx.cpp:
```diff
@@ -39,6 +40,40 @@ void RemovePrintOps(std::shared_ptr<Graph>& graph) {
   removePrintOps(graph->block());
 }
 
+Value* insertIValueWithoutTupleConstants(Graph* graph, const IValue& ival) {
```
I feel like this sets the trend that each new constant type will need a new onnx pass to remove it from the graph.  That seems like much more work than simply putting a rule into onnx so that it knows how to deal with the tuple constants directly.  Depending on the actual issue it might be a good idea to teach tuple lowering about tuple constants, so that they get lowered appropriately.

eellison replied:
`# onnx does not support tuples, so try to remove them` - I don't think onnx handles tuples at all. i think it fails if there are any tuples you can't lower in the graph. I added this pass to fix a test that was failing.

zdevito replied:
Maybe the test is failing because the LowerAllTuples pass doesn't correctly handle tuple constants. If so, it would be more concise to fix that pass, with the added benefit that we would be able to lower tuple constants there as well.

eellison replied:
Constant Propagation will already handle lowering of tuple constants

eellison replied:
When I implemented `str[]` and `float[]` constants, I communicated with onnx folks and they're adding support for exporting it. I don't think anytime someone adds constants to the JIT the burden is necessarily on them to also fix onnx representation, in this case it was an easy fix and less hassle so I implemented it.

zdevito reviewed (commented):
This looks pretty good, but I'd am not sure about cleaning up tuple constants in onnx as a pass.

eellison commented on torch/csrc/jit/constants.cpp:
```diff
@@ -16,6 +16,19 @@ c10::OperatorOptions aliasAnalysisInternalSpecialCase() {
+    const c10::optional<SourceRange>& loc,
+    const c10::optional<ScopePtr>& scope) {
+  for (const auto& elem : ivalue.toTuple()->elements()) {
+    if (tryInsertConstant(g, elem, loc, scope) == c10::nullopt) {
```
We will need to check for tensors that require gradient, but yes i can figure out a way to not add dead constants in the graph.

eellison commented on torch/csrc/jit/ir.cpp:
```diff
@@ -150,53 +248,21 @@ static void printTypeList(
-    case AttributeKind::ts:
-      out << "[<Tensors>]";
-      break;
+    PRINT_ATTRIBUTE(f)
```
I can write out the cases, i thought in this case a macro would be be easier to parse.

eellison commented on torch/csrc/jit/irparser.cpp:
```diff
@@ -213,38 +196,91 @@ void IRParser::parseAttr(Node* n) {
+ *   name = "Bob"
+ *   coefs = [1.2, 3.4, 0.6]
+ */
+void IRParser::parseAttr(Node* n) {
```
By IVale of 1 do you mean a tensor? Tensor is printed as `{1}` and int doesn't have the brackets. Tensors can't be parsed (before or after this PR). Yea sure, i can remove the `irparser` code. I know @ZolotukhinM & co were using it to write passes but they probably don't need tuple constants.

zdevito replied:
The ambiguity is `.ival_(foo, 1)` vs .i_(foo, i)`, I think both might print the same way, so they won't make it through serialization and parsing

eellison replied:
removed irparsing

Commit:
Update on "[JIT] implement tuple constants"


Add Tuple Constants to JIT. The constraint here is that all elements of a tuple must themself be insertable as a a constant. Previously tuples were special cased in constant propagation, but now that there are more passes that are inserted constants, such as freezing, we should just have tuples be representable as constants. 



[ghstack-poisoned]

Commit:
Update on "[JIT] implement tuple constants"


Add Tuple Constants to JIT. The constraint here is that all elements of a tuple must themself be insertable as a a constant. Previously tuples were special cased in constant propagation, but now that there are more passes that are inserted constants, such as freezing, we should just have tuples be representable as constants. 



[ghstack-poisoned]

Commit:
Update on "[JIT] implement tuple constants"


Add Tuple Constants to JIT. The constraint here is that all elements of a tuple must themself be insertable as a a constant. Previously tuples were special cased in constant propagation, but now that there are more passes that are inserted constants, such as freezing, we should just have tuples be representable as constants. 



[ghstack-poisoned]

eellison commented:
@zdevito updated to modify lower tuples pass, and I believe addressed other comments.

Commit:
Update on "[JIT] implement tuple constants"


Add Tuple Constants to JIT. The constraint here is that all elements of a tuple must themself be insertable as a a constant. Previously tuples were special cased in constant propagation, but now that there are more passes that are inserted constants, such as freezing, we should just have tuples be representable as constants. 



[ghstack-poisoned]

Commit:
Update on "[JIT] implement tuple constants"


Add Tuple Constants to JIT. The constraint here is that all elements of a tuple must themself be insertable as a a constant. Previously tuples were special cased in constant propagation, but now that there are more passes that are inserted constants, such as freezing, we should just have tuples be representable as constants. 



[ghstack-poisoned]

zdevito commented on torch/csrc/jit/ir.cpp:
```diff
@@ -136,6 +147,95 @@ static void printStrList(
+  }
+}
+
+static void printAttribute(
```
Why is this a function?

zdevito commented on torch/csrc/jit/ir.cpp:
```diff
@@ -136,6 +147,95 @@ static void printStrList(
+  out << "[<Tensors>]";
+}
+
+static void printAttribute(
```
Same here.

zdevito commented on torch/csrc/jit/ir.cpp:
```diff
@@ -96,34 +96,45 @@ std::ostream& operator<<(
+PRINT_LIST(c10::List, bool, item)
+// clang-format on
+
+static void printAttribute(std::ostream& out, const std::string& attr) {
```
Why is this a function, it just tail calls another function.

zdevito commented on torch/csrc/jit/ir.cpp:
```diff
@@ -136,6 +147,95 @@ static void printStrList(
+    printAttribute(out, ival.to##type()); \
+    return;                               \
+  };
+  PRINT_IVALUE(Int)
```
Can this not use repr? Why do we need new IValue printing code here?

zdevito commented on torch/csrc/jit/node_hashing.cpp:
```diff
@@ -40,6 +40,121 @@ bool typeListEqual(
+  if (a1.tagKind() != a2.tagKind()) {
+    return false;
+  }
+#define COMPARE_IVALUE(type)                              \
```
still seems to be a macro here

zdevito commented on torch/csrc/jit/passes/lower_tuples.cpp:
```diff
@@ -76,6 +77,41 @@ void removeTupleNodes(Node* n, bool must_remove_tuples) {
 
 static void LowerAllTuples(Block* block);
 
+static void removeTupleConstant(Node * n) {
```
I think this needs to be part of LowerAllTuples, because  right now LowerAllTuple doesn't lower all tuples, which is misleading.

Commit:
Update on "[JIT] implement tuple constants"


Add Tuple Constants to JIT. The constraint here is that all elements of a tuple must themself be insertable as a a constant. Previously tuples were special cased in constant propagation, but now that there are more passes that are inserted constants, such as freezing, we should just have tuples be representable as constants.

Differential Revision: [D19439514](https://our.internmc.facebook.com/intern/diff/D19439514)

[ghstack-poisoned]

Commit:
Update on "[JIT] implement tuple constants"


Add Tuple Constants to JIT. The constraint here is that all elements of a tuple must themself be insertable as a a constant. Previously tuples were special cased in constant propagation, but now that there are more passes that are inserted constants, such as freezing, we should just have tuples be representable as constants.

Differential Revision: [D19439514](https://our.internmc.facebook.com/intern/diff/D19439514)

[ghstack-poisoned]

Commit:
Update on "[JIT] implement tuple constants"


Add Tuple Constants to JIT. The constraint here is that all elements of a tuple must themself be insertable as a a constant. Previously tuples were special cased in constant propagation, but now that there are more passes that are inserted constants, such as freezing, we should just have tuples be representable as constants.

Differential Revision: [D19439514](https://our.internmc.facebook.com/intern/diff/D19439514)

[ghstack-poisoned]

eellison reviewed (commented):
okay, updated

Commit:
Update on "[JIT] implement tuple constants"


Add Tuple Constants to JIT. The constraint here is that all elements of a tuple must themself be insertable as a a constant. Previously tuples were special cased in constant propagation, but now that there are more passes that are inserted constants, such as freezing, we should just have tuples be representable as constants.

Differential Revision: [D19439514](https://our.internmc.facebook.com/intern/diff/D19439514)

[ghstack-poisoned]

driazati commented on torch/csrc/jit/attributes.h:
```diff
@@ -19,10 +19,21 @@ using ::c10::Symbol;
 constexpr int max_tensor_display_size = 10;
 
-enum class AttributeKind { f, fs, i, is, s, ss, t, ts, g, gs, ty, tys };
+enum class AttributeKind { f, fs, i, is, s, ss, t, ts, g, gs, ty, tys, ival };
```
nit: can you name it `ivalue` instead of `ival`? The abbreviations here are already pretty messy and unclear (like what is `tys`?)

eellison replied:
`ty` is type, and `tys` is list of `ty`.

driazati commented on torch/csrc/jit/constants.cpp:
```diff
@@ -16,6 +16,32 @@ c10::OperatorOptions aliasAnalysisInternalSpecialCase() {
+  if (ivalue.isTensor()) {
+    return insertableTensor(ivalue.toTensor());
+  }
+  if (ivalue.isList() || ivalue.isTuple()) {
```
Does this mean that lists can be implicitly converted to tuple constants?

eellison replied:
I don't know what you mean

driazati commented on torch/csrc/jit/constants.cpp:
```diff
@@ -16,6 +16,32 @@ c10::OperatorOptions aliasAnalysisInternalSpecialCase() {
 }
 } // namespace
 
+bool insertableTensor(const at::Tensor& ten) {
```
Shouldn't the [`val.isTensor`](https://github.com/pytorch/pytorch/pull/31841/files?file-filters%5B%5D=.cpp&file-filters%5B%5D=.h#diff-a28bcf9c5b0581cf3dd1179580c2f957R65) case for `tryInsertConstant` also have this check?

eellison replied:
maybe, but i think the stack trace would be worse in that case.

driazati commented on torch/csrc/jit/ir.cpp:
```diff
@@ -96,44 +96,42 @@ std::ostream& operator<<(
-  for (auto& item : items) {
-    if (i++ > 0) {
-      out << ", ";
+static void printAttribute(std::ostream& out, const at::Tensor& tensor) {
```
The previous conversation was resolved by why do we need `printAttribute` instead of `IValue`'s `<<` print operator/`repr`? Even the non-`AttributeKind::ival` values can just be boxed in a `IValue` before printing

eellison replied:
We're printing attributes in a specific way that is different than our general ivalue printing - the tensors `{Tensors}` thing

driazati commented on torch/csrc/jit/node_hashing.cpp:
```diff
@@ -19,27 +19,90 @@ bool tensorEqual(const at::Tensor& lhs, const at::Tensor& rhs) {
+}
+
+template <typename attribute_type> // int64_t, bool, double
+bool attributesEqual(attribute_type a1, attribute_type a2) {
```
maybe `static_assert` that `attribute_type` is one of the supported types so this doesn't get used accidentally where it shouldn't

eellison replied:
i have a assert failure lower in the file, and there it's commented of when this function is okay to call into

driazati commented on torch/csrc/jit/node_hashing.cpp:
```diff
@@ -19,27 +19,90 @@ bool tensorEqual(const at::Tensor& lhs, const at::Tensor& rhs) {
+    return true;
+  }
+  if (a1.isString()) {
+    return attributesEqual(a1.toStringRef(), a2.toStringRef());
```
is this not an infinite loop? there's no overload of `attributesEqual` for `std::string` as far as I can tell and the default just calls `ivaluesEqual`

eellison replied:
i think it resolves correctly because it doesn't infinite loop, but agreed it is confusing. will change.

Commit:
Update on "[JIT] implement tuple constants"


Add Tuple Constants to JIT. The constraint here is that all elements of a tuple must themself be insertable as a a constant. Previously tuples were special cased in constant propagation, but now that there are more passes that are inserted constants, such as freezing, we should just have tuples be representable as constants.

Differential Revision: [D19439514](https://our.internmc.facebook.com/intern/diff/D19439514)

[ghstack-poisoned]

Commit:
Update on "[JIT] implement tuple constants"


Add Tuple Constants to JIT. The constraint here is that all elements of a tuple must themself be insertable as a a constant. Previously tuples were special cased in constant propagation, but now that there are more passes that are inserted constants, such as freezing, we should just have tuples be representable as constants.

Differential Revision: [D19439514](https://our.internmc.facebook.com/intern/diff/D19439514)

[ghstack-poisoned]

Commit:
Update on "[JIT] implement tuple constants"


Add Tuple Constants to JIT. The constraint here is that all elements of a tuple must themself be insertable as a a constant. Previously tuples were special cased in constant propagation, but now that there are more passes that are inserted constants, such as freezing, we should just have tuples be representable as constants.

Differential Revision: [D19439514](https://our.internmc.facebook.com/intern/diff/D19439514)

[ghstack-poisoned]

Commit:
Update on "[JIT] implement tuple constants"


Add Tuple Constants to JIT. The constraint here is that all elements of a tuple must themself be insertable as a a constant. Previously tuples were special cased in constant propagation, but now that there are more passes that are inserted constants, such as freezing, we should just have tuples be representable as constants.

Differential Revision: [D19439514](https://our.internmc.facebook.com/intern/diff/D19439514)

[ghstack-poisoned]
